### ISSUE TITLE
CPU training is reproducible only on first epoch

### ISSUE THREAD

**Issue by yylonly** [keras-team/keras#10201]
I already set seed same as Keras FAQ https://keras.io/getting-started/faq/#how-can-i-obtain-reproducible-results-using-keras-during-development at the beginning of my code. Moreover, I set seed for the dense layer initializer.

First Model:
Epoch 1: [2.3224406, 0.10505]
Epoch 2: [2.012029, 0.44036666]

Second Model
Epoch 1: [2.3224406, 0.10505]
Epoch 2: [2.1158938, 0.31053334]

Code is shown as follows:

```
seed = 12345

import numpy as np
np.random.seed(seed)
import tensorflow as tf
import random as rn

# The below is necessary in Python 3.2.3 onwards to
# have reproducible behavior for certain hash-based operations.
# See these references for further details:
# https://docs.python.org/3.4/using/cmdline.html#envvar-PYTHONHASHSEED
# https://github.com/keras-team/keras/issues/2280#issuecomment-306959926

import os
os.environ['PYTHONHASHSEED'] = '0'

# The below is necessary for starting Numpy generated random numbers
# in a well-defined initial state.

np.random.seed(seed)

# The below is necessary for starting core Python generated random numbers
# in a well-defined state.

rn.seed(seed)

# Force TensorFlow to use single thread.
# Multiple threads are a potential source of
# non-reproducible results.
# For further details, see: https://stackoverflow.com/questions/42022950/which-seeds-have-to-be-set-where-to-realize-100-reproducibility-of-training-res

session_conf = tf.ConfigProto(intra_op_parallelism_threads=1, inter_op_parallelism_threads=1)

from keras import backend as K

# The below tf.set_random_seed() will make random number generation
# in the TensorFlow backend have a well-defined initial state.
# For further details, see: https://www.tensorflow.org/api_docs/python/tf/set_random_seed

tf.set_random_seed(seed)

sess = tf.Session(graph=tf.get_default_graph(), config=session_conf)
K.set_session(sess)

# Rest of code follows ...

from keras.layers import *
from keras.models import *
from keras.optimizers import *
from keras.datasets.mnist import load_data
from keras.utils.np_utils import to_categorical
from keras.initializers import glorot_uniform

(X_train, y_train), (X_test, y_test) = load_data()
y_train = to_categorical(y_train)
y_test = to_categorical(y_test)

XX = X_train
YY = y_train

adam = Adam(lr=0.001, beta_1=0.9, beta_2=0.999, epsilon=1e-08, decay=0)

model = Sequential()
model.add(Flatten(input_shape=X_train.shape[1:]))
model.add(Lambda(lambda x:x/255.0))
model.add(Dense(512, activation='relu', kernel_initializer=glorot_uniform(seed)))
#model.add(Dropout(0.2, seed=seed))
model.add(Dense(10, activation='softmax',  kernel_initializer=glorot_uniform(seed)))

model.compile(optimizer=adam, 
              loss='categorical_crossentropy',
              metrics=['accuracy'])

#for i in range(1):
loss = model.train_on_batch(X_train, y_train)
loss1 = model.train_on_batch(XX, YY)  
print(loss)
print(loss1)

#model.fit(X_train, y_train, batch_size=128, epochs=2, validation_data=(X_test, y_test), shuffle=False)

model = Sequential()
model.add(Flatten(input_shape=X_train.shape[1:]))
model.add(Lambda(lambda x:x/255.0))
model.add(Dense(512, activation='relu', kernel_initializer=glorot_uniform(seed)))
#model.add(Dropout(0.2, seed=seed))
model.add(Dense(10, activation='softmax',  kernel_initializer=glorot_uniform(seed)))

model.compile(optimizer=adam, 
              loss='categorical_crossentropy',
              metrics=['accuracy'])

loss = model.train_on_batch(X_train, y_train)
loss1 = model.train_on_batch(XX, YY)  
print(loss)
print(loss1)
#model.fit(X_train, y_train, batch_size=128, epochs=2, validation_data=(X_test, y_test), shuffle=False)
```

Any ideas? Thanks!

**Comment by yylonly** [comment]
Further investigation:
When change Adam to SGD without momentum, the result is reproducible.

**Comment by mrgkumar** [comment]
On training a Dense layer network (without dropouts) the results are reproducible even for Adam optimizer using CNTK backend.  I tried with several versions of tensorflow from 1.2.1 till 1.13.1  all of them have issues in reproducing the results.
